src/devices/funkbus.c

```diff
@@ -0,0 +1,197 @@
+     |     |     |  |  |     |  |  |
+    _|     |__ __|  |__|     |__|  |__.....
+     |  0  |  0  | 	1  |  0  |  1  |
+
+    mic: parity + unknown 4bit checksum
+    bits: 48
+    structure: see funkbus_packet_t
+
+
+    some details can be found by searching
+    "instafunk RX/TX-Modul pdf"
+
+*/
+
+#include "decoder.h"
+
+typedef enum {
+    FB_ACTION_STOP,
+    FB_ACTION_OFF,
+    FB_ACTION_ON,
+    FB_ACTION_SCENE
+} funkbus_action_t;
+
+typedef struct {
+    uint8_t typ : 4; // there are multible types
+    uint8_t subtyp : 4;
+
+    uint32_t sn : 20;
+
+    uint8_t r1 : 2;  // unknown
+    uint8_t bat : 1; // battery low
+    uint8_t r2 : 1;  // unknown
+
+    uint8_t sw : 3; // button on the remote
+    uint8_t ch : 2; // remote channel 0-2 are switches 3 == light scene
+    uint8_t r3 : 1; // unknown
+
+    funkbus_action_t action : 2;
+    uint8_t repeat : 1; // 1 == not first send of packet
+    uint8_t global : 1; // 1 == global ON / OFF button
+    uint8_t parity : 1; // parity over all bits
+    uint8_t check : 4;  // how to calculate is unknown
+} __attribute__((packed)) funkbus_packet_t;
+
+static uint64_t get_data_lsb(uint8_t const *bitrow, size_t start, uint8_t end)
+{
+    uint64_t result = 0;
+    uint64_t mask   = 1;
+    result          = 0;
+    for (; start <= end; mask <<= 1)
+        if (bitrow_get_bit(bitrow, start++) != 0)
+            result |= mask;
+    return result;
+}
+
+static uint8_t calc_parity(uint8_t const *bitrow, size_t len)
+{
+    uint8_t result = 0;
+    for (uint8_t i = 0; i < len; i++) {
+        if (bitrow_get_bit(bitrow, i)) {
+            result++;
+        }
+    }
+    return result & 0x01;
+}
+
+static int funkbus_callback(r_device *decoder, bitbuffer_t *bitbuffer)
+{
+
+    for (uint16_t row = 0; row < bitbuffer->num_rows; row++) {
+        uint16_t len = bitbuffer->bits_per_row[row];
+        uint8_t *bin  = bitbuffer->bb[row];
+
+        if (len < 48) {
+            return DECODE_ABORT_LENGTH;
+        }
+
+        uint8_t get_c = 0;
+#define get(x) \
+    get_data_lsb(bin, get_c, get_c + x - 1); \
+    get_c += x
+
+        funkbus_packet_t packet;
+        packet.typ    = get(4);
+        packet.subtyp = get(4);
+
+        // remote
+        if (packet.typ != 0x4 || packet.subtyp != 0x3) {
+            return DECODE_ABORT_EARLY;
+        }
+
+        packet.sn     = get(20);
+        packet.r1     = get(2);
+        packet.bat    = get(1);
+        packet.r2     = get(2);
+        packet.sw     = get(3);
+        packet.ch     = get(2);
+        packet.r3     = get(1);
+        packet.action = get(2);
+        packet.repeat = get(1);
+        packet.global = get(1);
+        packet.parity = get(1);
+        packet.check  = get(4); // how to calc?
+
+        if (packet.parity != calc_parity(bin, 43)) {
+            return DECODE_FAIL_MIC;
+        }
+
+        // calc id based on sn, ch + sw
+        uint64_t id = packet.sw + (packet.ch << 4) + (packet.sn << 8);
```

Comment:
Isn't the SN a unique ID?

Reply:
each remote I have has a different serial number.
is the `id` unique per function / channel or per device in the context of rtl_433?

Reply:
it should be per (pyhsical) sender device, that way we can collect sent data per-sender.
If you need a key per function, e.g. in MQTT you can use a event template of `[/model][/id][/channel]`

Reply:
setting `id` == `sn` is fin for me, the connection to my automation is not written yet ;)

`channel` in context of the funkbus is more a Layer of the remote.
for example this remote here has a A / B / C button which changes the the "layer"
![image](https://user-images.githubusercontent.com/974709/144719049-9ec22776-a6f4-43b6-a444-915b1068ae5e.png)

does that mean that `channel` for rtl == `channel` + `sw` for the funkbus?

Reply:
"channel" is used as a more persistent kind of ID, mainly on devices with a random ID a battery change.
If it's influenced by the user on keypress then I'd say layer is a good description, or maybe "target" or "dest". (if I understand correctly that A/B/C is not to distinguish senders but to select receivers, right?)
I'll have to browse our decoders to see what's commonly used. This would be a good addition to our "common keys" docs.

Reply:
the funkbus system is not hardwired to a specific target.
each receivers can be set in to a "learning mode" where codes can be added or removed by long pressing a button on the remote.
on button can control none, one or more receivers.

in the most cases each button in a layer will control a different receiver e.g power plug adapter or dimming light.

for example:
if `A` ist pressed and after 1+ then
the remote will send a code for Layer / Group A channel 1 ON

if `A` ist pressed and after 2- then
the remote will send a code for Layer / Group A channel 2 OFF

if `B` ist pressed and after 1+ then
the remote will send a code for Layer / Group B channel 1 ON
 
 the groups are mapped like
 A --> 0
 B --> 1
 C --> 2
 the value 3 is special since its used for scenes (the rounds buttons on the lower half)
 there the Layer / Group selection does not have an effect at all.